src/wirecloud/commons/search.py

```diff
@@ -0,0 +1,365 @@
+                "keyword_lowercase": {
+                    "tokenizer": "keyword",
+                    "filter": ["lowercase"]
+                }
+            },
+            "tokenizer": {
+                "ngram_tokenizer": {
+                    "type": "ngram",
+                    "min_gram": 2,
+                    "max_gram": 20,
+                    "token_chars": ["letter", "digit"]
+                }
+            }
+        }
+    },
+    "mappings": {
+        "properties": {
+            "name": {
+                "type": "text",
+                "analyzer": "ngram_analyzer",
+                "search_analyzer": "standard",
+                "fields": {
+                    "keyword": {"type": "keyword"}
+                }
+            }
+        }
+    }
+}
+
+_available_search_engines = None
+_available_rebuild_engines = None
+
+
+class UserOut(BaseModel):
+    fullname: str
+    username: str
+    # TODO: Organizations
+
+
+class GroupOut(BaseModel):
+    name: str
+    # TODO: Organizations
+
+
+class SearchResponse(BaseModel):
+    offset: int
+    pagecount: int
+    pagelen: int
+    pagenum: int
+    results: list[Union[UserOut, GroupOut, WorkspaceOutResponse, ResourceOutResponse]]
+    total: int
+
+
+def get_available_search_engines() -> dict[str, Callable[[str, int, int, str | None], SearchResponse]]:
+    global _available_search_engines
+
+    if _available_search_engines is None:
+        from src.wirecloud.platform.search import search_workspaces
+        from src.wirecloud.catalogue.search import search_resources
+        _available_search_engines = {"group": search_groups,
+                                     "user": search_users,
+                                     "workspace": search_workspaces,
+                                     "usergroup": search_user_groups,
+                                     "resource": search_resources}
+
+    return _available_search_engines
+
+
+def get_available_rebuild_engines() -> dict[str, Callable[[DBSession], None]]:
+    global _available_rebuild_engines
+
+    if _available_rebuild_engines is None:
+        from src.wirecloud.platform.search import rebuild_workspace_index
+        from src.wirecloud.catalogue.search import rebuild_resource_index
+        _available_rebuild_engines = {"group": rebuild_group_index,
+                                      "user": rebuild_user_index,
+                                      "workspace": rebuild_workspace_index,
+                                      "resource": rebuild_resource_index,
+                                      "all": rebuild_all_indexes}
+
+    return _available_rebuild_engines
+
+
+def is_available_search_engine(indexname: str) -> bool:
+    return indexname in get_available_search_engines()
+
+
+def is_available_rebuild_engine(indexname: str) -> bool:
+    return indexname in get_available_rebuild_engines()
+
+
+def get_search_engine(indexname: str) -> Optional[Callable[[str, int, int, str | None], SearchResponse]]:
```

Comment:
Cuidado! La sintaxis `str | None` es de python 9 y tratamos de soportar python 8. Cámbialo por `Union[str, None]`, o aun mejor `Optional[str]`.

```suggestion
def get_search_engine(indexname: str) -> Optional[Callable[[str, int, int, Optional[str]], SearchResponse]]:
```